cpp/src/arrow/type_test.cc

```diff
@@ -364,152 +365,344 @@ TEST(TestField, TestMerge) {
+}
+template <>
+auto&& FieldPathTestCase::GetInput<ArrayData>() const {
+  return this->array->data();
+}
+template <>
+auto&& FieldPathTestCase::GetInput<ChunkedArray>() const {
+  return this->chunked_array;
+}
+template <>
+auto&& FieldPathTestCase::GetInput<RecordBatch>() const {
+  return this->record_batch;
+}
+template <>
+auto&& FieldPathTestCase::GetInput<Table>() const {
+  return this->table;
+}
 
-  EXPECT_TRUE(column0->Equals(v0));
-  EXPECT_TRUE(column0->Equals(v2_0));
+template <>
+auto&& FieldPathTestCase::OutputValues::Get<Field>() const {
+  return this->field;
+}
+template <>
+auto&& FieldPathTestCase::OutputValues::Get<Array>() const {
+  return this->array;
+}
+template <>
+auto&& FieldPathTestCase::OutputValues::Get<ArrayData>() const {
+  return this->array->data();
+}
+template <>
+auto&& FieldPathTestCase::OutputValues::Get<ChunkedArray>() const {
+  return this->chunked_array;
+}
 
-  EXPECT_TRUE(column1->Equals(v1));
-  EXPECT_TRUE(column1->Equals(v2_1_0));
-  EXPECT_FALSE(column1->Equals(v2_1));
+class FieldPathTestFixture : public ::testing::Test {
+ public:
+  FieldPathTestFixture() : case_(FieldPathTestCase::Instance()) {}
 
-  EXPECT_TRUE(column2_1->Equals(v2_1));
+ protected:
+  template <typename T>
+  using OutputType = typename internal::FieldPathGetType<T>::element_type;
 
-  EXPECT_RAISES_WITH_MESSAGE_THAT(NotImplemented,
-                                  HasSubstr("Get child data of non-struct chunked array"),
-                                  FieldPath({0}).Get(*column0));
-}
+  template <typename I>
+  void AssertOutputsEqual(const std::shared_ptr<Field>& expected,
+                          const std::shared_ptr<Field>& actual) const {
+    AssertFieldEqual(*expected, *actual);
+  }
+  template <typename I>
+  void AssertOutputsEqual(const std::shared_ptr<Array>& expected,
+                          const std::shared_ptr<Array>& actual) const {
+    AssertArraysEqual(*expected, *actual);
+  }
+  template <typename I>
+  void AssertOutputsEqual(const std::shared_ptr<ChunkedArray>& expected,
+                          const std::shared_ptr<ChunkedArray>& actual) const {
+    if constexpr (std::is_same_v<I, ChunkedArray>) {
+      EXPECT_EQ(case_->chunked_array->num_chunks(), actual->num_chunks());
+    } else {
+      EXPECT_EQ(expected->num_chunks(), actual->num_chunks());
+    }
+    AssertChunkedEquivalent(*expected, *actual);
+  }
+
+  const FieldPathTestCase* case_;
+};
+
+class TestFieldPath : public FieldPathTestFixture {
+ protected:
+  template <typename I, bool Flattened = false>
+  void TestGetWithInvalidIndex() const {
+    const auto& input = case_->GetInput<I>();
+    for (const auto& path :
+         {FieldPath({2, 1, 0}), FieldPath({1, 2, 0}), FieldPath{1, 1, 2}}) {
+      EXPECT_RAISES_WITH_MESSAGE_THAT(IndexError,
+                                      ::testing::HasSubstr("index out of range"),
+                                      internal::GetChild<Flattened>(*input, path));
```

Comment:
Since the `IndexError` has a more detailed error message, is it possible to test it precisely at least for a hard-coded case?